lib/apiservers/engine/backends/archive.go

```diff
@@ -487,13 +490,92 @@ func (rm *ArchiveStreamReaderMap) FindArchiveReaders(containerSourcePath string)
 		}
 
 		if !found {
-			nodes = append(nodes, startingNode)
+			// prepend the starting node at the beginning
+			nodes = append([]*ArchiveReader{startingNode}, nodes...)
 		}
+	} else {
+		startingNode = nodes[0]
+	}
+
+	err = rm.buildFilterSpec(containerSourcePath, nodes, startingNode)
+	if err != nil {
+		return nil, err
 	}
 
 	return nodes, nil
 }
 
+func (rm *ArchiveStreamReaderMap) buildFilterSpec(containerSourcePath string, nodes []*ArchiveReader, startingNode *ArchiveReader) error {
+	var err error
+
+	// Build an exclusion filter for each writer.  For example, the reader for / should not read
+	// from submounts as there are separate readers for those.
+	buildExclusion := func(path string, node *ArchiveReader) error {
+		var walkResults []string
+
+		childWalker := func(prefix patricia.Prefix, item patricia.Item) error {
+			if _, ok := item.(*ArchiveReader); !ok {
+				return fmt.Errorf("item not ArchiveReader")
+			}
+
+			ar, _ := item.(*ArchiveReader)
+			walkResults = append(walkResults, ar.mountPoint.Destination)
+			return nil
+		}
+
+		// prefix = current node's mount path
+		nodePrefix := patricia.Prefix(path)
+
+		err = rm.prefixTrie.VisitSubtree(nodePrefix, childWalker)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to build exclusion filter for %s: %s", path, err.Error())
+			log.Error(msg)
+			return fmt.Errorf(msg)
+		}
+
+		for _, child := range walkResults {
+			if child != path {
+				node.filterSpec.Exclusions[child] = struct{}{}
+			}
+		}
+
+		return nil
+	}
+
+	for _, node := range nodes {
+		// Clear out existing exclusions and inclusions
+		node.filterSpec.Exclusions = make(map[string]struct{})
+		node.filterSpec.Inclusions = make(map[string]struct{})
+
+		err = buildExclusion(node.mountPoint.Destination, node)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Add inclusion filter.  When there is an inclusion, there should be only one node in
+	// the slice that is returned.
+	//
+	//	Example 1:
+	//		containerSourcePath -	/file.txt
+	//
+	//		ArchiveReader path -	/
+	//		Inclusion filter -		file.txt
+	//
+	//	Example 2:
+	//		containerSourcePath -	/mnt/A/a/file.txt
+	//
+	//		ArchiveReader path -	/mnt/A
+	//		Inclusion filter -		a/file.txt
+	inclusionPath := strings.TrimPrefix(containerSourcePath, startingNode.mountPoint.Destination)
+	inclusionPath = strings.TrimPrefix(inclusionPath, "/")
+	if len(nodes) == 1 {
+		nodes[0].filterSpec.Inclusions[inclusionPath] = struct{}{}
+	}
+
```

Comment:
I think we need Rebase for the correct packing/unpacking of the archive:
```
node.filterSpec.Rebase = startingNode.mountPoint.Destination
```

unpack
```
for file := range archive {
  stripped := strings.TrimPrefix(file.Path, filter.Rebase)
  if included && !exlcuded {
     // write file to target filesystem at filesystem_mountpoint/stripped
  }
}
```

pack
```
for file := range FilesUnderMountpoint {
  if included && !exlcuded {
     stripped := strings.TrimPrefix(file.Path, filesystem-mountpoint)
     rebased := path.Join(filter.Rebase, stripped)
     // add file to archive with header path "rebased"
  }
}
```